larq_compute_engine/mlir/transforms/prepare_tf.cc

```diff
@@ -25,11 +25,41 @@ DenseElementsAttr GetConstantVector(Attribute filter, float val) {
   return DenseElementsAttr::get(type, val);
 }
 
-bool IsBinaryFilter(Attribute filter) {
-  if (!filter.isa<DenseElementsAttr>()) return false;
+DenseElementsAttr GetScaleVector(Attribute filter_attr) {
+  auto filter = filter_attr.cast<DenseElementsAttr>();
+  auto filter_type = filter_attr.getType().cast<ShapedType>();
+  auto channels = filter_type.getShape()[3];
+  auto element_type = filter_type.getElementType();
+
+  std::vector<Attribute> scales(channels);
+  for (std::size_t i = 0; i < channels; ++i) {
+    auto scale = std::abs(filter.getValue<float>({0, 0, 0, i}));
+    scales[i] = FloatAttr::get(element_type, scale);
+  }
+
+  RankedTensorType type = RankedTensorType::get({channels}, element_type);
+  return DenseElementsAttr::get(type, scales);
+}
 
-  for (auto value : filter.cast<DenseElementsAttr>().getValues<float>()) {
-    if (std::abs((std::abs(value) - 1.0f)) > 0.005f) return false;
+bool IsBinaryFilter(Attribute filter_attr) {
+  if (!filter_attr.isa<DenseElementsAttr>()) return false;
+  auto filter = filter_attr.cast<DenseElementsAttr>();
+
+  auto shape = filter_attr.getType().cast<ShapedType>().getShape();
+  if (shape.size() != 4) return false;
+
+  std::size_t h, w, i, o;
+  for (h = 0; h < shape[0]; ++h) {
+    for (w = 0; w < shape[1]; ++w) {
+      for (i = 0; i < shape[2]; ++i) {
+        for (o = 0; o < shape[3]; ++o) {
+          auto scale = filter.getValue<float>({0, 0, 0, o});
+          if (scale == 0.0f) return false;
```

Comment:
This check is only to make sure that the division below is always well defined. Is this really necessary in that case?

Reply:
[Apparently](https://stackoverflow.com/questions/12114498/can-a-near-zero-floating-value-cause-a-divide-by-zero-error) dividing by zero does not crash but simply gives infinity (or NaN for `0/0`), but it can also happens for numbers really close to zero. So to be safe and not output models with infinity/NaN, i think we can add it (maybe with a smaller tolerance like `1e-10f` or so).

Reply:
Makes sense. I changed it in a8d774b4e52bb06cac554fc05a270f9b6f86db35

We can also do the other computations in double if we really want to be more precise at some point.